crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopInvariant.hs

```diff
@@ -0,0 +1,1151 @@
+                ?logMessage $ "SimpleLoopInvariant: SymbolicBranchState: ComputeFixpoint"
+
+                loc <- W4.getCurrentProgramLoc sym
+                C.addAssumption bak $ C.BranchCondition loc (C.pausedLoc inside_loop_frame) loop_condition
+
+                C.ExecutionFeatureNewState <$>
+                  runReaderT
+                    (C.resumeFrame (C.forgetPostdomFrame inside_loop_frame) $ sim_state ^. (C.stateTree . C.actContext))
+                    sim_state
+
+              CheckFixpoint _fixpoint_record -> do
+                ?logMessage $ "SimpleLoopInvariant: SymbolicBranchState: CheckFixpoint"
+
+                return C.ExecutionFeatureNoChange
+
+      _ -> return C.ExecutionFeatureNoChange
+
+
+advanceFixpointState ::
+  forall sym bak ext p rtp blocks r args .
+  (C.IsSymBackend sym bak, C.HasLLVMAnn sym, ?memOpts :: C.MemOptions, ?logMessage :: String -> IO ()) =>
+  bak ->
+  C.GlobalVar C.Mem ->
+  (InvariantPhase -> [Some (W4.SymExpr sym)] -> MapF (W4.SymExpr sym) (InvariantEntry sym) -> IO (W4.Pred sym)) ->
+  C.BlockID blocks args ->
+  C.SimState p sym ext rtp (C.CrucibleLang blocks r) ('Just args) ->
+  FixpointState p sym ext rtp blocks ->
+  IORef (FixpointState p sym ext rtp blocks) ->
+  IO (C.ExecutionFeatureResult p sym ext rtp)
+
+advanceFixpointState bak mem_var loop_invariant block_id sim_state fixpoint_state fixpoint_state_ref = do
+  let ?ptrWidth = knownNat @64
+  let sym = C.backendGetSym bak
+  loc <- W4.getCurrentProgramLoc sym
+  case fixpoint_state of
+    BeforeFixpoint -> do
+      ?logMessage $ "SimpleLoopInvariant: RunningState: BeforeFixpoint -> ComputeFixpoint " ++ show block_id ++ " " ++ show (pretty (W4.plSourceLoc loc))
+      assumption_frame_identifier <- C.pushAssumptionFrame bak
+      let mem_impl = fromJust $ C.lookupGlobal mem_var (sim_state ^. C.stateGlobals)
+      let res_mem_impl = mem_impl { C.memImplHeap = C.pushStackFrameMem "fix" $ C.memImplHeap mem_impl }
+
+--      ?logMessage $ "SimpleLoopInvariant: start memory\n" ++ (show (C.ppMem (C.memImplHeap mem_impl)))
+
+      -- Get a fresh block value that doesn't correspond to any valid memory region
+      havoc_blk <- W4.natLit sym =<< C.nextBlock (C.memImplBlockSource mem_impl)
+
+      writeIORef fixpoint_state_ref $ ComputeFixpoint $
+        FixpointRecord
+        { fixpointBlockId = block_id
+        , fixpointAssumptionFrameIdentifier = assumption_frame_identifier
+        , fixpointSubstitution = VarSubst MapF.empty MapF.empty
+        , fixpointRegMap = sim_state ^. (C.stateCrucibleFrame . C.frameRegs)
+        , fixpointMemSubstitution = MemSubst mempty
+        , fixpointInitialSimState = sim_state
+        , fixpointImplicitParams = []
+        , fixpointHavocBlock = havoc_blk
+        }
+      return $ C.ExecutionFeatureModifiedState $ C.RunningState (C.RunBlockStart block_id) $
+        sim_state & C.stateGlobals %~ C.insertGlobal mem_var res_mem_impl
+
+    ComputeFixpoint fixpoint_record
```

Comment:
This is a particularly large case. It might be worth breaking this up into some helper functions to make it more readable.